### ISSUE TITLE
Use redux?

### ISSUE THREAD

**Issue by rolyatmax** [rolyatmax/redux-resource-manager#12]
Do we really _need_ to use redux with this library? We don't really need the flexibility to add more and more actions and our state mutations are pretty simple. Perhaps we could get away with not using redux at all. If users _want_ this library to fire `RESOURCE_FETCH`, `RESOURCE_RECEIVED`, or `RESOURCE_ERROR` actions, we could provide a straightforward way for them to subscribe to these types of events. I think we'd need to provide some sort of subscription or hook for users to use to cause a re-render in their application anyway.

Furthermore, this `connect-to-redux` functionality could also be a little smarter about when we trigger a "store change" event. Currently, any update to a resource (even ones not currently used by any components) will cause a redux store mutation and trigger a re-render. But if we didn't use redux under the hood, we could be a little smarter, running the `mapResourcesToProps` function and comparing the output to the previous output and only triggering a "store change" if the outputs differed.

Also, anything we can do to limit our dependencies and the technologies that users would have to adopt to use this library is a win.

Definitely curious to hear others' thoughts.

**Comment by mannynotfound** [comment]
Seems like the most useful part of this library is 

```
    const { user } = props;

    // user <-- This is a resource object. It has a few special properties:
    // user.status <-- this is the status of the request. It can be one of these three values:
    //                 ('pending' || 'fulfilled' || 'rejected')
    // user.retry <-- if a resource's status is 'rejected', this will be a function that a component
    //                can use to retry the request
    // user.result <-- if the resource's status is 'fulfilled', this will be the value returned from
    //                 the api

    if (user.status === 'pending') return <Loader />;
    if (user.status === 'rejected') return <RetryButton retry={user.retry} />;

   // render etc
}
```

The redux part would be handy if you needed to trigger actions in other parts of your app based on API loading state where a global store with the state could give you that information, otherwise usually just the components care about having the data.

I could see this just being a high-order component with a built in place for a "loader" and some settings for retrying, caching, timeouts, error handling, etc (all the crap no one wants to re-invent). Said high-order component could take redux actions (or any func) as props and fire them on loading events if needed, otherwise you can just use it as a pure react component.

Interesting library that's doing similar things: [react-resolver](https://github.com/ericclemmons/react-resolver)

**Comment by rolyatmax** [comment]
Looking through `react-resolver` thanks to your rec!

> The redux part would be handy if you needed to trigger actions in other parts of your app based on API loading state where a global store with the state could give you that information

I think this is where the "subscription" part might be able to help out. Would giving users a way to fire actions on `RESOURCE_FETCH`, `RESOURCE_RECEIVED`, and `RESOURCE_ERROR` be sufficient you think? Maybe you could optionally pass in arbitrary callbacks when creating the resource manager?:

``` js
createResourceManager({
  onFetch: (requestParams) => dispatch({ type: 'FETCH', requestParams }),
  onReceived: (requestParams, data) => dispatch({ type: 'RECEIVED', requestParams, data }),
  onError: (requestParams, error, retry) => dispatch({ type: 'ERROR', requestParams, error, retry })
})
```

Or if we wanted to be able to optionally connect the resource manager to a redux store for some smarter integrations, maybe the API is simply:

``` js
connectResourceManagerToStore(resourceManager, store)

// or

connectToStore(resourceManager)(store)
```

> I could see this just being a high-order component with a built in place for a "loader" and some settings for retrying, caching, timeouts, error handling, etc (all the crap no one wants to re-invent). 

The `DataRequired` component that we use internally on the LT team at BF basically does all this (with the exception of timeouts). I've struggled with whether or not it should just come as part of [the higher order component](https://github.com/rolyatmax/redux-resource-manager/blob/master/src/connect_resource_manager.js#L12) that `connectResourceManager` wraps a component with. I found that there were times when we wouldn't necessarily want to show a loader - and times when we did want to show a loader but _only_ until a certain subset of the resources had returned. It ended up being useful for the components to have the option of examining `resource.status` themselves and doing something custom if it was `pending` or `rejected`.

If we _did_ decide to combine the loader/error-showing functionality of `DataRequired` with the higher order component we wrap the Component with in `connectResourceManager`, perhaps it could take a number of options that would (a) pass the resources through to the child component unwrapped (keeping their fields: `status`, `result`, `retry`, etc), (b) show a custom `Loader` or `Error` component, (c) show these components or children when certain resources have loaded (this could just be a `shouldShowLoader` function supplied by the developer that receives the resources as args and expects a boolean return value).

Also, the `connectResourceManager` could easily take the form of a decorator (which results in a syntax I quite like!).

Whew - that was a lot of stuff. Did any of that make sense? Hahaha

**Comment by rolyatmax** [comment]
One other piece of functionality I like about the resource manager is that definitions for the resources can live in one place - even if you end up using that resource in several unrelated components. I think there's something powerful about this. Since the resources definition really has more to do with the endpoints than with the client application itself, these could be shared across many applications which use the same APIs.

For example, imagine if we just needed to maintain a single `resources.js` definition for all of the internal endpoints at BF. Applications could just name the `bf-resources` definition as a dependency and start using `getResource.users()` or `getResource.videos()` right away. You could do the same for other public APIs.

Also, also: I like that even if multiple components are wrapped with this `connectResourceManager`, they still use a common cache. I didn't see this functionality in `react-resolver`, though it wouldn't be too difficult for the developer to implement some sort of cache in the function passed to `@resolve`.

Also, also, also: I'm loving `react-resolver`'s docs!

**Comment by mannynotfound** [comment]
Agreed that it seems useful to define all your resources in one place that can be shared across a teams / apps. Kind of like [CSS Modules](https://github.com/css-modules/webpack-demo/blob/master/src/components/0-Logo/Logo.js) but for data. Taking everything mentioned above maybe something like

``` js
import {manage} from 'resource-manager'
import resources from 'bf-resources';
import Loader from './Loader';
import Error from './Error';

@manage("video", function(props) {
  const totallyOptionalConfig = {
    params: {videoId: props.params.videoId},
    request: { }, 
    cache: '5m',
    onFetch: () => {},
    onReceived: () => {},
    onError: () => {},
  };

  return resources.video(totallyOptionalConfig);
});
class VideoDetail extends React.Component {
  render() {
    const { video } = this.props;
    // this could be handled by the component however they wish
    // but maybe provide an optional wrapper that does this in a standardized way
    if (video.status === 'loading') return <Loader />
    if (video.status === 'error') return <Error />
    if (video.result) return <Video video={video.result} />
    ...
  }
}
```

**Comment by rolyatmax** [comment]
Ohhh, I'm liking this!

A few thoughts:
1. In order to share a cache across multiple components wrapped with `manage`, I think we'd have to have a `createResourceManager` that would take the `resources` definition (and perhaps some options for "connecting" to a Redux store) and would return the `manage` function we could use to wrap components.
2. Do you think the cache TTL should be defined per component? Or just per resource? If just per resource, this could be included in the `resources` definition. FWIW, this is how we're doing it now.
3. We could let `manage` take some options for defining `Loader` and `Error` components so it could do the job of checking a resource's `status`.
4. Maybe by default `manage` would pass down just the resource's `result`, but could also take an option to have it pass down a resource as is (with `status`, `result`, `retry`, etc).

Working off what you've got, maybe it would become something like this?:

In `manager.js`:

``` js
import createResourceManager from 'resource-manager'
import resources from 'bf-resources'
import store from './store'
import { RESOURCE_FETCH, RESOURCE_RECEIVED, RESOURCE_ERROR } from './actions'


const { manage } = createResourceManager(resources, {
  onFetch: (params) => store.dispatch({ type: RESOURCE_FETCH, params }),
  onReceived: (params, data) => store.dispatch({ type: RESOURCE_FETCH, params, data }),
  onError: (params, error, retry) => store.dispatch({ type: RESOURCE_FETCH, params, error, retry })
})

export default {
  manage
}
```

Meanwhile, in the Component:

``` js
import { manage } from './manager'
import Loader from './Loader'
import ErrorBar from './Error'

@manage(
  (props, getResources) => ({
    video: getResources.video({ videoId: props.videoId })
  }),
  {
    LoaderComponent: Loader,
    ErrorComponent: ErrorBar
  }
)
class VideoDetail extends React.Component {
  render() {
    // video comes unwrapped by default (i.e. it is just the value of video.result)
    const { video } = this.props
    // when given a LoaderComponent & ErrorComponent, the `manage` wrapper takes
    // care of `pending` and `rejected` statuses, and unwraps and passes the result
    // when the status is `fulfilled`
    return <Video video={video} />
  }
}
```

**Comment by mannynotfound** [comment]
> Do you think the cache TTL should be defined per component? Or just per resource? If just per resource, this could be included in the resources definition. FWIW, this is how we're doing it now.

Yeah I was just firing off things I think you would want to "configure" like cache ttl or request options, but it probably makes more sense to just bundle those options with the resource. 

I like the above proposed spec, esp since you could call `createResourceManager` without the 2nd arg if you werent using redux. Abstracting out the "load status" to be handled by the manager seems nice but probably makes it harder to do custom rendering based on _how many_ resources have loaded or which specific ones have loaded. Probably more edge case than norm but im just thinking how you do it so you can easily create progress loaders or just not block render if some resources load before others.

**Comment by rolyatmax** [comment]
> Abstracting out the "load status" to be handled by the manager seems nice but probably makes it harder to do custom rendering based on how many resources have loaded or which specific ones have loaded. Probably more edge case than norm but im just thinking how you do it so you can easily create progress loaders or just not block render if some resources load before others.

Yeah, I think you're right. If we were to include the "load status" handling in the `manager`, we'd have to also provide some hooks whereby developers could pass a `shouldShowLoader` function. I'm also realizing that even if we didn't provide those hooks, developers could still choose to have the `manager` pass through the full request objects untouched so they could handle them however they want. You could also create a higher order component to wrap the `manager` component to encapsulate your custom Loader-showing logic.
